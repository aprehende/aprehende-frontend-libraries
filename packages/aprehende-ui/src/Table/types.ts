import React from 'react';

type ColsByName<Data> = {
  [key in keyof Data]: {
    key: keyof Data;
    title: string;
    render?: (
      value: Data[key],
      index: number
    ) => React.ReactNode | boolean | '' | null;
    isActions?: boolean;
  };
};

type Column<TData> = ColsByName<TData>[keyof ColsByName<TData>];

export interface TableProps<T> {
  data: T[];
  columns: Column<T>[];
  widthFirstRow?: string;
  isSticky?: boolean;
}
