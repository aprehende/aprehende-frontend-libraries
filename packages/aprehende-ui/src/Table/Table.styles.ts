import styled from '@emotion/styled';

export const THead = styled.thead<{ isSticky?: boolean; top: string }>`
  width: 100%;
  box-sizing: border-box;
  position: ${({ isSticky }) => isSticky && 'sticky'};
  top: ${({ top }) => top};
`;

export const THeader = styled.tr`
  width: 100%;
  flex: none;
  order: 0;
  flex-grow: 0;
  display: flex;
  cursor: pointer;
  align-items: center;
  background: #dfe6e9;
  align-self: stretch;
  border-radius: 4px 4px 0 0;
  justify-content: space-between;
`;

export const TableStyled = styled.table`
  width: 100%;
  display: flex;
  border-radius: 4px;
  box-sizing: border-box;
  flex-direction: column;
  align-items: flex-start;
  border: 1px solid #cbd5e1;
`;

export const TH = styled.th<{ customWidth?: string; isActions?: boolean }>`
  width: 100%;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  padding: 10px;

  text-align: left;
  ${({ isActions }) =>
    isActions &&
    `display: flex;
    justify-content: space-between;
    align-items: center;
    max-width: 70px;
    padding: 0;
    `}
`;

export const ContainerRow = styled.tr`
  display: flex;
  align-items: center;
  width: 100%;
  box-sizing: border-box;
  justify-content: space-between;
  text-align: center;
`;

export const TD = styled.td<{ customWidth?: string; isActions?: boolean }>`
  width: 100%;
  padding: 10px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  box-sizing: border-box;
  text-align: left;
  ${({ isActions }) =>
    isActions &&
    `display: flex;
    justify-content: space-between;
    align-items: center;
    max-width: 70px;
    padding: 0;
    `}
`;

export const TBody = styled.tbody`
  width: 100%;
  overflow-y: auto;
  overflow-x: auto;
  box-sizing: border-box;
`;
