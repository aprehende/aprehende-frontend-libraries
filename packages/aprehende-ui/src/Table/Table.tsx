import { useEffect, useRef, useState } from 'react';

import {
  ContainerRow,
  TableStyled,
  TBody,
  TD,
  TH,
  THead,
  THeader,
} from './Table.styles';
import { TableProps } from './types';

function Table<TableData>({
  data,
  columns,
  widthFirstRow,
  isSticky,
}: TableProps<TableData>) {
  const tableRef = useRef<HTMLTableElement>(null);

  const [top, setTop] = useState<string>('0px');

  useEffect(() => {
    setTop(`${tableRef.current?.getBoundingClientRect().top}px`);
  }, []);

  return (
    <TableStyled ref={tableRef}>
      <THead isSticky={isSticky} top={top}>
        <THeader>
          {columns.map((column, idx) => (
            <TH
              customWidth={widthFirstRow}
              key={idx}
              isActions={column.isActions}
            >
              {column.title}
            </TH>
          ))}
        </THeader>
      </THead>
      <TBody>
        {data.map((row, idx) => (
          <ContainerRow key={idx}>
            {columns.map(({ key, render, isActions }, idxColumn) => (
              <TD
                customWidth={widthFirstRow}
                key={`${idx}-${idxColumn}`}
                isActions={isActions}
              >
                {!render && (row[key] as string)}
                {render && render(row[key], idx)}
              </TD>
            ))}
          </ContainerRow>
        ))}
      </TBody>
    </TableStyled>
  );
}

export default Table;
