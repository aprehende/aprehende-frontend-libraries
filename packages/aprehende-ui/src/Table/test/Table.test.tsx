import { render } from '@testing-library/react';
import { describe, it } from 'vitest';

import Table from '../Table';

interface User {
  id: number;
  name: string;
  email: string;
  properties: {
    city: string;
    country: string;
  };
}

const users: User[] = [
  {
    id: 1,
    name: 'John Doe',
    email: 'johndoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
  {
    id: 2,
    name: 'Jane Doe',
    email: 'janedoe@mail.com',
    properties: {
      city: 'New York',
      country: 'USA',
    },
  },
];

describe('Table', () => {
  it('should render', () => {
    const { asFragment } = render(
      <Table<User>
        data={users}
        columns={[
          {
            title: 'Name',
            key: 'name',
            render: (name: string) => <strong>{name}</strong>,
          },
          {
            title: 'Email',
            key: 'email',
            render: (email: string) => <a href={`mailto:${email}`}>{email}</a>,
          },
          {
            title: 'City',
            key: 'properties',
            render: (properties: User['properties']) => properties.city,
          },
          {
            title: 'Country',
            key: 'properties',
            render: (properties: User['properties']) => properties.country,
          },
        ]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the data', () => {
    const { getByText } = render(
      <Table<User>
        data={users}
        columns={[
          {
            title: 'Name',
            key: 'name',
            render: (name: string) => <strong>{name}</strong>,
          },
          {
            title: 'Email',
            key: 'email',
            render: (email: string) => <a href={`mailto:${email}`}>{email}</a>,
          },
          {
            title: 'City',
            key: 'properties',
            render: (properties: User['properties']) => properties.city,
          },
          {
            title: 'Country',
            key: 'properties',
            render: (properties: User['properties']) => properties.country,
          },
        ]}
      />
    );
    expect(getByText('John Doe')).toBeInTheDocument();
  });

  it('should render sticky table', () => {
    const { getByTestId } = render(
      <Table<User>
        isSticky
        data={users}
        columns={[
          {
            title: 'Name',
            key: 'name',
            render: (name: string) => <strong>{name}</strong>,
          },
          {
            title: 'Email',
            key: 'email',
            render: (email: string) => <a href={`mailto:${email}`}>{email}</a>,
          },
          {
            title: 'City',
            key: 'properties',
            render: (properties: User['properties']) => properties.city,
          },
          {
            title: 'Country',
            key: 'properties',
            render: (properties: User['properties']) => properties.country,
          },
        ]}
      />
    );
    const tableHeader = getByTestId('thead');
    expect(tableHeader).toHaveStyle('position: sticky');
  });

  it('should render table with actions', () => {
    render(
      <Table<User>
        data={users}
        columns={[
          {
            title: 'Name',
            key: 'name',
            render: (name: string) => <strong>{name}</strong>,
          },
          {
            title: 'Email',
            key: 'email',
          },
          {
            title: 'City',
            key: 'properties',
            render: (properties: User['properties']) => properties.city,
          },
          {
            title: 'Country',
            key: 'properties',
            render: (properties: User['properties']) => properties.country,
          },
          {
            title: 'Actions',
            key: 'id',
            isActions: true,
            render: () => (
              <span>
                <button>edit</button>
              </span>
            ),
          },
        ]}
      />
    );
  });
});
