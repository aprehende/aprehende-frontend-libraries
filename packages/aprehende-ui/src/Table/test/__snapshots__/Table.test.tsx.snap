// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should render 1`] = `
<DocumentFragment>
  <table
    class="css-179tqhh"
    role="table"
  >
    <thead
      class="css-hqx5pz"
      data-testid="thead"
    >
      <tr
        class="css-1o8h9rs"
      >
        <th
          class="css-1y8wkbc"
        >
          Name
        </th>
        <th
          class="css-1y8wkbc"
        >
          Email
        </th>
        <th
          class="css-1y8wkbc"
        >
          City
        </th>
        <th
          class="css-1y8wkbc"
        >
          Country
        </th>
      </tr>
    </thead>
    <tbody
      class="css-fefzs"
      data-testid="tbody"
    >
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            John Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:johndoe@mail.com"
          >
            johndoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
      <tr
        class="css-1qnhu6f"
        data-testid="tr"
      >
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <strong>
            Jane Doe
          </strong>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          <a
            href="mailto:janedoe@mail.com"
          >
            janedoe@mail.com
          </a>
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          New York
        </td>
        <td
          class="css-16yn0fw"
          data-testid="td"
        >
          USA
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
